src/restricted-endpoints/replication/changes/changes.controller.spec.ts

```diff
@@ -112,18 +111,85 @@ describe('ChangesController', () => {
     ]);
   });
 
-  it('should not return the document content', async () => {
-    const res = await getChangesResults([
-      { subject: 'School', action: 'read' },
-    ]);
+  it('should not return the document content on default', async () => {
+    getRulesSpy.mockReturnValue([{ subject: 'School', action: 'read' }]);
+
+    const res = await controller.changes('some-db', user);
 
     res.results.forEach((r) => expect(r.doc).toBeUndefined());
   });
 
-  function getChangesResults(rules: DocumentRule[], params?) {
-    jest.spyOn(mockRulesService, 'getRulesForUser').mockReturnValue(rules);
-    jest.spyOn(mockCouchdbService, 'get').mockReturnValue(of({ ...changes }));
+  it('should return the document content if requested', async () => {
+    getRulesSpy.mockReturnValue([{ subject: 'School', action: 'read' }]);
+
+    const res = await controller.changes('some-db', user, {
+      include_docs: 'true',
+    });
+
+    res.results.forEach((r) => expect(r.doc).toBeDefined());
+  });
+
+  it('should keep requesting docs until the limit is reached', async () => {
+    getRulesSpy.mockReturnValue([{ subject: 'Child', action: 'read' }]);
+    getSpy
+      .mockReturnValueOnce(of(createChanges([schoolDoc, privateSchoolDoc], 2)))
+      .mockReturnValueOnce(of(createChanges([childDoc, deletedChildDoc])));
+
+    const res = await controller.changes('some-db', user, { limit: 2 });
+
+    expect(res.pending).toBe(0);
+    expect(res.last_seq).toBe(docToChange(deletedChildDoc).seq);
+    expect(res.results.map((r) => r.id)).toEqual([
+      childDoc._id,
+      deletedChildDoc._id,
+    ]);
+  });
+
+  it('should not return more changes than requested', async () => {
+    getRulesSpy.mockReturnValue([{ subject: 'Child', action: 'read' }]);
+    getSpy
+      .mockReturnValueOnce(
+        of(createChanges([schoolDoc, childDoc, childDoc], 2)),
+      )
+      .mockReturnValueOnce(of(createChanges([schoolDoc, childDoc, childDoc])));
+
+    const res = await controller.changes('some-db', user, { limit: 3 });
+
+    expect(res.pending).toBe(1);
+    expect(res.last_seq).toBe(docToChange(childDoc).seq);
+    res.results.forEach(({ id }) => expect(id).toBe(childDoc._id));
```

Comment:
I don't fully understand this test case: if the request is `limit: 3`, why is it getting something with "pending" then?
Maybe I am confused with the two mocked responses and reused items. Does this mean the total changes (global) are [schoolDoc, childDoc, childDoc, schoolDoc, childDoc, childDoc], i.e. length 6?

Reply:
Yes the global changes are correct. So the user requests 3 changes and can only read children docs and therefore receives`[childDoc, childDoc, childDoc]`. In the database there is one more change after the last `childDoc` that will not be returned so there is still one pending change.

Reply:
But your right, the `pending 2` was incorrect in the first changes feed.